haskell/test/Model/BaseDAO/Types.hs

```diff
@@ -0,0 +1,236 @@
+-- SPDX-FileCopyrightText: 2021 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# OPTIONS_GHC -Wno-missing-signatures #-}
+
+module Model.BaseDAO.Types
+  ( ModelT (..)
+  , ModelState (..)
+  , SimpleContract (..)
+  , SimpleOperation (..)
+  , runModelT
+  , getStore
+  , getConfig
+  , modifyStore
+  , addOperations
+  , getEntrypoint
+  , mkSimpleOperation
+  , (-$$?)
+
+  , ModelCall (..)
+  , ModelError (..)
+  , contractErrorToModelError
+  )where
+
+import Universum
+
+import Control.Monad.Except (MonadError, throwError)
+import Data.Typeable (cast)
+import qualified Data.Map as M
+import Fmt
+
+import Michelson.Typed.Value
+import Michelson.Interpret
+import Michelson.Text
+import qualified Michelson.Typed as T
+import Lorentz hiding (not, cast, get)
+import Tezos.Address (ContractHash(..))
+import Tezos.Core
+import qualified Lorentz.Contracts.Spec.FA2Interface as FA2
+
+import Ligo.BaseDAO.Types
+
+-- | Transformer used in haskell implementation of BaseDAO
+newtype ModelT a = ModelT
+  { unModelT :: (ExceptT ModelError $ State ModelState) a
+  } deriving newtype (Functor, Applicative, Monad, MonadState ModelState, MonadError ModelError)
+
+-- | A type that keep track of blockchain state
+data ModelState = ModelState
+  { msFullStorage :: FullStorage
+  , msOperations :: [SimpleOperation]
+
+  , msLevel :: Natural
+  , msChainId :: ChainId
+  , msSelfAddress :: Address
+  , msGovernanceTokenAddress :: Address
+  , msContracts :: Map (Address, Text) SimpleContract
```

Comment:
I'm a bit confused by this too and `SimpleContract` as well :thinking: 

Aren't we only interested in contract that:
- have an FA2 transfer entrypoint
- can receive a transfer
- aren't either of the above

If so, I'd think that `SimpleContract` should be a sum-type and the key of the map only be an `Address`.

Reply:
Yeah, that would be a better abstraction, I'll make the changes accordingly.

Reply:
Updated this.